bee-common/bee-packable-derive/src/trait_impl.rs

```diff
@@ -71,19 +75,35 @@ impl TraitImpl {
                         #unpack
                     }));
 
-                    tag_decls.push(quote!(const #tag_ident: #tag_type = #tag;))
+                    tag_decls.push(quote!(const #tag_ident: #tag_type = #tag;));
+
+                    tag_idents_and_variants.push((tag_ident, variant_ident));
+                }
+
+                let mut tag_asserts = Vec::with_capacity(len * (len - 1) / 2);
+
+                for (i, (fst, fst_variant)) in tag_idents_and_variants.iter().enumerate() {
+                    if let Some(idents_and_variants) = tag_idents_and_variants.get((i + 1)..) {
+                        for (snd, snd_variant) in idents_and_variants {
+                            let tag_assert = quote!(
+                                const _: () = assert!(#fst != #snd, concat!("The tags for the variants `", stringify!(#fst_variant), "` and `", stringify!(#snd_variant) ,"` of enum `", stringify!(#enum_ident), "` are equal"));
+                            );
+                            tag_asserts.push(tag_assert);
+                        }
+                    }
                 }
+
                 Ok(Self {
-                    ident: input.ident,
+                    ident: enum_ident.clone(),
                     generics: input.generics,
                     unpack_error,
                     pack: quote!(match self {
                         #(#pack_arms)*
                     }),
                     unpack: quote! {
                         #(#tag_decls)*
+                        #(#tag_asserts)*
```

Comment:
Out of curiosity, why is this in unpack ?

Reply:
It is because we cannot generate code like:
```rust
match tag {
    expr_for_tag_0 => {...}
    expr_for_tag_1 => {...}
    expr_for_tag_2 => {...}
    ...
}
```
this is because not every expression is a valid pattern. Then we have to declare a bunch of constants beforehand:
```rust
const CONST_IDENT_0: TagType := expr_for_tag_0;
const CONST_IDENT_1: TagType := expr_for_tag_1;
const CONST_IDENT_2: TagType := expr_for_tag_2;
...

match tag {
    CONST_IDENT_0 => {...}
    CONST_IDENT_1 => {...}
    CONST_IDENT_2 => {...}
    ...
}
```

so this is in unpack because is the only method where the tag expressions are assigned to an identifier and we can write things like:
```rust
const _: () = if CONST_IDENT_0 == CONST_IDENT_1 {
    panic!("here be dragons")
}
```
Which I hope saves us some constant evaluation time instead of doing:
```rust
const _: () = if expr_for_tag_0 == expr_for_tag_1 {
    panic!("here be dragons")
}
```